src/Common/ThreadStatus.h

```diff
@@ -57,45 +57,57 @@ using ThreadStatusPtr = ThreadStatus *;
   * Create via CurrentThread::initializeQuery (for queries) or directly (for various background tasks).
   * Use via CurrentThread::getGroup.
   */
+class ThreadGroupStatus;
+using ThreadGroupStatusPtr = std::shared_ptr<ThreadGroupStatus>;
+
 class ThreadGroupStatus
 {
 public:
-    struct ProfileEventsCountersAndMemory
-    {
-        ProfileEvents::Counters::Snapshot counters;
-        Int64 memory_usage;
-        UInt64 thread_id;
-    };
-
-    mutable std::mutex mutex;
+    /// The first thread created this thread group
+    UInt64 master_thread_id = 0;
 
     ProfileEvents::Counters performance_counters{VariableContext::Process};
     MemoryTracker memory_tracker{VariableContext::Process};
 
-    ContextWeakPtr query_context;
-    ContextWeakPtr global_context;
+    /// Access to the members below has to be in critical section with  mutex
+    mutable std::mutex mutex;
 
     InternalTextLogsQueueWeakPtr logs_queue_ptr;
     InternalProfileEventsQueueWeakPtr profile_queue_ptr;
-    std::function<void()> fatal_error_callback;
-
-    std::unordered_set<UInt64> thread_ids;
-    std::unordered_set<ThreadStatusPtr> threads;
-
-    /// The first thread created this thread group
-    UInt64 master_thread_id = 0;
 
     LogsLevel client_logs_level = LogsLevel::none;
 
     String query;
     UInt64 normalized_query_hash = 0;
```

Comment:
Let's make a getter and setter for this as well?
Also, we don't need a setter for `normalized_query_hash`. We can calculate it into `query` setter.

Another option is to make a separate structure for mutable stuff and make a common getter that returns this struct + lock.

Reply:
I thought about that. Seemed like a big step from totally public struct to such techniques. But If you are encouraging such approach lets do it!